tools/runtest

```diff
@@ -909,90 +1020,97 @@ fi
 #  !---------------------------------------------------------------------!
 #  ! Run tests                                                           !
 #  !---------------------------------------------------------------------!
-
 cd "$testdir"
-
-
 for buildtype in $buildtypes; do
-
   cdate=`date +'%m/%d/%Y'`
   ctime=`date +'%r'`
 
   echo "" | tee -a $QUICK_HOME/.quick.${buildtype}.runtest.log
   echo  "Testing $QUICK_VERSION $buildtype version started on $cdate at $ctime. " | tee -a $QUICK_HOME/.quick.${buildtype}.runtest.log
   echo "" | tee -a $QUICK_HOME/.quick.${buildtype}.runtest.log
-  
 
   # set qexe and testlist
-  set_qexe_testlist
-
-  # initialize the total number of test cases
-  total_tests=0 
+  set_qexe_testlist "$buildtype"
 
   # get the number of tests to run
-  get_total_tests
+  get_total_tests "$test_ene" "$test_grad" "$test_opt" "$test_api"
 
-  tc=1 # test count
   cd "$testdir/runs/$buildtype"
 
+  if [ $has_parallel = 'yes' ]; then
+    # ensure no leftover lock files from failed previous runs
+    rm -f "$QUICK_HOME/.quick.${buildtype}.runtest.log.lock" "$QUICK_HOME/.quick.${buildtype}.runtest.log.deleteme"
+  fi
+
   # Run energy tests
+  tc_start=0
+  tc_end=0
+
   if [ "$test_ene" = 'yes' ]; then
-    for t in `awk '{print $1}' "$test_list"| grep "ene"`; do
-      set_test
-      run_test
-      if [ `check_f_error` -eq 0 ]; then
-          check_energy
-          check_dipole
-          clean_up
-         
-          echo ""
-      fi
-      tc=$((tc+1))
-    done
+    tc_start=1
+    tc_end="$tot_ene_tests"
+    if [ $has_parallel = 'no' ]; then
+      for tc in `seq $tc_start $tc_end`; do
+        run_energy_test "${tests[$((tc-1))]}" "${test_names[$((tc-1))]}" "$tc" "$total_tests" \
+		"$testdir" "$QUICK_HOME/.quick.${buildtype}.runtest.log"
+      done
+    elif [ $has_parallel = 'yes' ]; then
+      parallel --keep-order --group --jobs "$PARALLEL_TEST_COUNT" --will-cite run_energy_test {1} {2} {3} \
+	      "$total_tests" "$testdir" "$QUICK_HOME/.quick.${buildtype}.runtest.log" {%} \
+	      ::: "${tests[@]:$((tc_start-1)):$tc_end}" :::+ "${test_names[@]:$((tc_start-1)):$tc_end}" \
```

Comment:
Older parallel versions do not provide `:::+`. If I change this into `:::` then the tests run (no hang) on chinotto (CentOS 7.9) but produce errors.

Reply:
Replacing `:::+` with `--xapply` in combination with `:::` works.